src/CoreWCF.Primitives/src/CoreWCF/Security/SecuritySessionServerSettings.cs

```diff
@@ -1956,6 +1968,570 @@ public Task OpenAsync(CancellationToken token)
+                        if (!_isInputClosed)
+                        {
+                            _isInputClosed = true;
+                            setInputSessionCloseHandle = true;
+
+                            if (!_isOutputClosed)
+                            {
+                                _closeRequestContext = requestContext;
+                                // CreateCloseResponse closes the message passed in
+                                _closeResponseMessage = CreateCloseResponse(message, null, token);
+                                cleanupContext = false;
+                            }
+                        }
+                    }
+
+                    if (setInputSessionCloseHandle)
+                    {
+                       _inputSessionCloseHandle.Set();
+                    }
+                    if (cleanupContext)
+                    {
+                        requestContext.CloseAsync();
+                        cleanupContext = false;
+                    }
+                }
+                finally
+                {
+                    message.Close();
+                    if (cleanupContext)
+                    {
+                        requestContext.Abort();
+                    }
+                }
+            }
+
+            protected override void OnCloseResponseMessageReceived(RequestContext requestContext, Message message, SecurityProtocolCorrelationState correlationState, TimeSpan timeout)
+            {
+                bool cleanupContext = true;
+                try
+                {
+                    bool isCloseResponseExpected = false;
+                    bool setInputSessionCloseHandle = false;
+                    lock (LocalLock)
+                    {
+                        isCloseResponseExpected = _sentClose;
+                        if (isCloseResponseExpected && !_isInputClosed)
+                        {
+                            _isInputClosed = true;
+                            setInputSessionCloseHandle = true;
+                        }
+                    }
+                    if (!isCloseResponseExpected)
+                    {
+                        _duplexCommObj.Fault(new ProtocolException(SR.Format(SR.UnexpectedSecuritySessionCloseResponse)));
+                        return;
+                    }
+                    if (setInputSessionCloseHandle)
+                    {
+                        _inputSessionCloseHandle.Set();
+                    }
+
+                    requestContext.CloseAsync();
+                    cleanupContext = false;
+                }
+                finally
+                {
+                    message.Close();
+                    if (cleanupContext)
+                    {
+                        requestContext.Abort();
+                    }
+                }
+            }
+
+            internal bool WaitForOutputSessionClose(TimeSpan timeout, out bool wasAborted)
+            {
```

Comment:
The caller of this method is async, so make this method async. I see that this method has an out which will need to be removed to make it async. You can use tuple return values like this:
```c#
            internal Task<(bool success, bool wasAborted)> WaitForOutputSessionCloseAsync(TimeSpan timeout)
            {
                wasAborted = false;
                try
                {
                    return (await _outputSessionCloseHandle.WaitAsync(timeout, false), false);
                }
                catch (CommunicationObjectAbortedException)
                {
                    if (State != CommunicationState.Closed) throw;
                    return (true, true);
                }
            }
```
Then at the call point you can do this:
```c#
bool didOutputSessionClose;
(didOutputSessionClose, wasAborted) = await WaitForOutputSessionCloseAsync(ServiceDefaults.CloseTimeout, out wasAborted);
```